### ISSUE TITLE
NLL: error from URL crate

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#47703]
In https://github.com/rust-lang/rust/issues/47596, @SimonSapin reported several NLL errors in dependencies of the servo crate. @Aaron1011 later minimized one of those errors into [this example](https://play.rust-lang.org/?gist=0bd31f0d318c0a4889f720fe54d7979d&version=undefined):

```rust
#![feature(nll)]

struct MyStruct<'a> {
    field: &'a mut (),
    field2: WithDrop
}

struct WithDrop;

impl Drop for WithDrop {

    fn drop(&mut self) {}

}

impl<'a> MyStruct<'a> {

    fn consume(self) -> &'a mut () { self.field }
}

fn main() {}
```

which yields:

```
error[E0597]: `*self.field` does not live long enough
  --> src/main.rs:18:38
   |
18 |     fn consume(self) -> &'a mut () { self.field }
   |                                      ^^^^^^^^^^  - borrowed value only lives until here
   |                                      |
   |                                      borrowed value does not live long enough
   |
   = note: borrowed value must be valid for lifetime '_#3r...
```

Removing feature(nll) makes the code work, so *probably* this is an NLL bug.

**Comment by SimonSapin** [comment]
As noted in the other issue, changing the method to the one below works around the bug:

```rust
    fn consume(self) -> &'a mut () {
        let MyStruct { field, field2 } = self;
        field
    }
```

In other words, when `self` is owned `self.field` should move the field out rather than (re-)borrow it even if that field’s type happens to be `&mut _`. If other fields of `self` have destructors, they should be dropped individually as if the struct had been deconstructed with a pattern.

**Comment by nikomatsakis** [comment]
Another related case, from `atomic_refcell`, minimized by @lqd [here](https://github.com/rust-lang/rust/issues/47596#issuecomment-360125480):

https://play.rust-lang.org/?gist=5b3934b08dec4d034dc5b00fba8647c3&version=nightly

**Comment by davidtwco** [comment]
I'll take a look at this one.

**Comment by davidtwco** [comment]
@nikomatsakis I would appreciate some pointers on where I should be looking for this, couldn't spot anything when I took a look at what was happening in the areas of the code that I'm familiar with. Not worked with the NLL code much.

**Comment by nikomatsakis** [comment]
@davidtwco I don't know that this is an NLL problem *per se*, I think it has more to do with MIR borrowck. In particular, removing `#![feature(nll)]` and compiling with `-Zborrowck=mir` still gives roughly the same error. I tried running with `-Zdump-mir=nll` to dump out the MIR, and I see that the heart of the MIR in the relevant file (`mir_dump/rustc.{{impl}}[1]-consume.-------.nll.0.mir`) is:

```
        StorageLive(_2);
        _2 = &mut (*(_1.0: &mut ()));
        _0 = &mut (*_2);
        StorageDead(_2);
```

I think the error comes precisely at `StorageDead(_2)`. I presume this is because of the borrow `0 = &mut *_2`, which is still in effect. However, it's not supposed to be an error to kill the storage for `_2` when `*_2` is borrowed (because `*_2` is a deref of a pointer; if `_2.x` were borrowed, that would be an error).

**Comment by nikomatsakis** [comment]
Tracing through, I expect that we start at a `StorageDead` item:

https://github.com/rust-lang/rust/blob/70f7d5842f29d4900f24420b030f144d21f3c5fc/src/librustc_mir/borrow_check/mod.rs#L424-L431

This invokes `access_place`:

https://github.com/rust-lang/rust/blob/70f7d5842f29d4900f24420b030f144d21f3c5fc/src/librustc_mir/borrow_check/mod.rs#L711-L724

which then checks for conflicts:

https://github.com/rust-lang/rust/blob/70f7d5842f29d4900f24420b030f144d21f3c5fc/src/librustc_mir/borrow_check/mod.rs#L744-L745

Inside the `check_access_for_conflict` action:

https://github.com/rust-lang/rust/blob/70f7d5842f29d4900f24420b030f144d21f3c5fc/src/librustc_mir/borrow_check/mod.rs#L753-L760

I expect we head down this arm:

https://github.com/rust-lang/rust/blob/70f7d5842f29d4900f24420b030f144d21f3c5fc/src/librustc_mir/borrow_check/mod.rs#L823

and then presumably here:

https://github.com/rust-lang/rust/blob/70f7d5842f29d4900f24420b030f144d21f3c5fc/src/librustc_mir/borrow_check/mod.rs#L859-L867

Stepping back, I *think* the fundamental problem comes from the call to `each_borrow_involving_path` in `check_access_for_conflict`. In particular, I think it ought to be screening out a borrow of `*foo`, since the `sd` parameter is "shallow" (which is supposed to mean: do not trace through references):

https://github.com/rust-lang/rust/blob/70f7d5842f29d4900f24420b030f144d21f3c5fc/src/librustc_mir/borrow_check/mod.rs#L762-L766

Looking at the source of that function:

https://github.com/rust-lang/rust/blob/70f7d5842f29d4900f24420b030f144d21f3c5fc/src/librustc_mir/borrow_check/mod.rs#L2054-L2061

I expect the problem must be in the `places_conflict` function invoked here:

https://github.com/rust-lang/rust/blob/70f7d5842f29d4900f24420b030f144d21f3c5fc/src/librustc_mir/borrow_check/mod.rs#L2076

But it *seems* like that function is returning `false` when we find a deref:

https://github.com/rust-lang/rust/blob/70f7d5842f29d4900f24420b030f144d21f3c5fc/src/librustc_mir/borrow_check/mod.rs#L1967-L1973

So I'm not sure where it goes wrong. Perhaps insert some debugs and see where my predictions go astray? =)

**Comment by nikomatsakis** [comment]
Ok, after some discussion with @davidtwco it has become clear where I went wrong. The error is actually being reported by a `Drop(_1)` terminator, where `_1` is the `self` path. The problem here is that we currently assume that `Drop(_1)` may access *any* data reachable from `_1` -- and that includes `_1.field`. If the type `MyStruct` had a destructor, that would be the right thing, and the error would be legit. But -- in this example -- that assumption is overly conservative.

I think what we have to do is to kind of "bake in" a bit of knowledge about the drop glue. For simplicity, I'll just describe what we need to do in terms of structs here as a kind of pseudo-code. 

Right now we do this:

```
check_drop(X: Place):
    Check that deep access to X is legal
```

This is too conservative. We want to be doing something like:

```
check_drop(X: Place):
    let T = typeof(X);
    if T is a struct S and S does not have a destructor {
        // Since S does not have a destructor, the default drop glue
        // will just unroll its fields and drop them one by one. So let's
        // check that all of those accesses will be legal.
        for each field F defined in S {
            check_drop(X.F) // i.e., extend the Place X with an access to the field F
        }
    } else if T needs drop {
        check that deep access to X is legal
    } else {
        // no code executes when T is dropped, so no checks needed
    }
}  
```

If we did this, we would wind up unrolling the `Drop(_1)` into a drop of `_1.field` and `_1.field2`. the type of `_1.field` is an `&mut`, and those do not need drop, so nothing happens there. The type of `_1.field2` is a struct with a dtor, so we'd check for conflicts on *that* path (and see none).
